### PR TITLE
Fixes for GitHub issues 964 and 1084

### PR DESCRIPTION
**Pull request by andreeis**
Fix for GitHub issue https://github.com/microsoft/vscode-cmake-tools/issues/1084:
Remove the default values for cmake.prefferedGenerators in package.json and use them later, only if the deduction process is not identifying any. Without this fix, Ninja and Unix Makefiles would always be preferred (when installed), when the user would not override these defaults in the settings and regardless of the preferred generator defined for the kit.

Fix for GitHub issue https://github.com/microsoft/vscode-cmake-tools/issues/964:
Always generate the VS kits with full information about the platforms, no defaults.
VS architecture becomes host architecture and is the arch before "_".
Target architecture is the arch after "_". 
If no "_" is present in the end of the kit name, then the one architecture applies for host and target.

I am investigating some test failures (probably only baseline changes) but I am submitting the PR before, to get results also from the CI.

### CODE REVIEW

File: src/kit.ts [hunk 1]
```diff
@@ -526,11 +561,31 @@ async function collectDevBatVars(devbat: string, args: string[], major_version:
 
 /**
  * Platform arguments for VS Generators
+ * For example, x86 and amd64_x86 will generate -A win32
  */
-const VsArchitectures: {[key: string]: string} = {
+const genPlatformFromVsHostTargetArchs: {[key: string]: string} = {
+  x86: 'win32',
```
Comment (bobbrow) [inline src/kit.ts:567]: If the key is a string, it would be best to add the keys with string syntax:

```ts
  'x86': 'win32',
```

File: src/kit.ts [hunk 2]
```diff
@@ -642,8 +703,11 @@ export async function scanForVSKits(pr?: ProgressReporter): Promise<Kit[]> {
   const installs = await vsInstallations();
   const prs = installs.map(async(inst): Promise<Kit[]> => {
     const ret = [] as Kit[];
-    const arches = ['x86', 'amd64', 'x86_amd64', 'x86_arm', 'amd64_arm', 'amd64_x86'];
-    const sub_prs = arches.map(arch => tryCreateNewVCEnvironment(inst, arch, pr));
+    // TODO: exclude toolsets that are not yet installed.
```
Comment (bobbrow) [inline src/kit.ts:706]: Doesn't `tryCreateNewVCEnvironment` handle your TODO?  I thought it would only succeed if the tools were there and skip it otherwise.

File: .vscode/launch.json [hunk 3]
```diff
@@ -213,18 +213,18 @@
             "request": "launch",
             "runtimeExecutable": "${execPath}",
             "args": [
-                "${workspaceFolder}/test/extension-tests/vs-preferred-gen/project-folder",
+                "${workspaceFolder}/test/extension-tests/successful-build/project-folder",
```
[the remark this answers is not in the document]
Comment (andreeis) [reply .vscode/launch.json:216]: Right. The original path was not working, there's no vs-preferred-gen in the CMake Tools test repo. But indeed the above change is redundant. So I'd rather delete the whole block instead than revert because it wasn't doing anything before.

File: src/kit.ts (hunk 2 above)
Comment (andreeis) [reply src/kit.ts:706]: It appears that vcvarsall.bat, when invoked for arm/arm64 on an install that does not include these architectures, it still creates a valid environment for x86. To me this is confusing, but I can follow up separately with the VS team, in case there is a good reason for this. The environment difference between the two cases is that for arm(64) installed, the first path is [hostArch]\[targetArch] instead of [hostArch]\[hostArch]. I am fixing it now by checking on this detail.

File: src/kit.ts (hunk 1 above)
Comment (andreeis) [reply src/kit.ts:567]: I tried but I get "Unnecessarily quoted property 'x86' found."

File: src/kit.ts [hunk 4]
```diff
@@ -434,6 +434,41 @@ function kitName(inst: VSInstallation, arch: string): string {
...
+  return hostArch;
+}
+
+/**
+ * Extract the toolset host architecture from the kit name suffix.
+ * The rule is: [hostArch]_[targetArch] or both [arch] if no "_" is present.
+ * @param hostTargetArch
+ */
+function kitTargetArch(hostTargetArch: string): string {
+  const targetArch: string = hostTargetArch.split("_")[1] || hostTargetArch;
```
[the remark this answers is not in the document]
Comment (andreeis) [reply src/kit.ts:453]: Yes, doing this for the next commit/push in this PR.

File: src/kit.ts [hunk 5]
```diff
@@ -461,12 +455,7 @@ function kitTargetArch(hostTargetArch: string): string {
  * @param targetArch The architecture of the target
  */
 function kitHostTargetArch(hostArch: string, targetArch: string): string {
-  let hostTargetArch: string = hostArch;
-  if (targetArch && vsHostTargetArchsFromGenPlatform[targetArch] !== hostArch) {
-    hostTargetArch = hostTargetArch.concat("_").concat(vsHostTargetArchsFromGenPlatform[targetArch]);
-  }
-
-  return hostTargetArch;
+  return hostArch === targetArch ? hostArch : hostArch.concat("_").concat(targetArch);
```
Comment (bobbrow) [inline src/kit.ts:458]: The second case is probably better as:
```ts
`${hostArch}_${targetArch}`
```
Using backtick quotes will evaluate the variables.

File: src/kit.ts [hunk 6]
```diff
@@ -602,18 +571,38 @@ const VsGenerators: {[key: string]: string} = {
...
   }
-  const variables = await collectDevBatVars(devbat, [`${arch}`], majorVersion, common_dir);
+  const variables = await collectDevBatVars(devbat, [`${kitHostTargetArch(hostArch, targetArch)}`], majorVersion, common_dir);
   if (!variables) {
     return null;
   } else {
+    // When invoked for arm or arm64, the vcvarsall.bat script creates an x86 environment,
+    // when the arm(64) components are not installed.
+    // The way to differentiate between the two cases is to check that the first path
+    // in the PATH environment variable ends with the target architecture.
```
Comment (bobbrow) [inline src/kit.ts:593]: Are you 100% sure this will always be the case forever?  I would hate to be inadvertently broken because msvc decided to change something without us knowing.

File: src/kit.ts [hunk 7]
```diff
@@ -602,18 +571,38 @@ const VsGenerators: {[key: string]: string} = {
   16: 'Visual Studio 16 2019'
 };
 
-async function varsForVSInstallation(inst: VSInstallation, arch: string): Promise<Map<string, string>|null> {
-  console.log(`varsForVSInstallation path:'${inst.installationPath}' version:${inst.installationVersion} arch:${arch}`);
+async function varsForVSInstallation(inst: VSInstallation, hostArch: string, targetArch: string): Promise<Map<string, string>|null> {
+  console.log(`varsForVSInstallation path:'${inst.installationPath}' version:${inst.installationVersion} host arch:${hostArch} - target arch:${targetArch}`);
+
+  // TODO: if needed, cover also the scenario of VS older than 2017,
```
Comment (bobbrow) [inline src/kit.ts:577]: I thought the older VS versions are handled on line 583.  It looks like it's searching the path you mention in this comment.  Can this TODO be removed?

Comment (andreeis) [reply src/kit.ts:577]: Oh I didn't realize that. It's true. I will remove the comment and also do another fix related to x64 and amd64, because the older vcvars scripts don't recognize x64.

File: src/kit.ts (hunk 6 above)
Comment (andreeis) [reply src/kit.ts:593]: No, not 100% sure. But I just noticed that a VS with arm/arm64 installed has some arm/arm64 specific vcvars, which call into the main one. I think that is a better indication.

**Comment by bobbrow** [comment]
Please double-check my merge.  There was a conflict with another feature that was recently submitted.

File: package.json [hunk 8]
```diff
@@ -1272,21 +1269,21 @@
     "mocha": "^4.1.0",
     "module-alias": "^2.2.1",
     "sinon": "~5.0.7",
-    "ts-loader": "^6.0.4",
-    "ts-node": "^6.0.0",
+    "ts-loader": "^6.2.2",
+    "ts-node": "^8.8.1",
```
Comment (bobbrow) [inline package.json:1273]: This new version of `ts-node` seems to be causing the build to fail on linux.  Can you please fix it or revert the change?

Comment (bobbrow) [reply package.json:1273]: I saw your update, but the build still fails.  Unfortunately, I don't think this was fixed.

Comment (andreeis) [reply package.json:1273]: Yes I just saw it fails. I am trying reverting next.
